source/common/network/listener_impl.cc

```diff
@@ -19,31 +19,41 @@
 namespace Envoy {
 namespace Network {
 
+Address::InstanceConstSharedPtr ListenerImpl::getLocalAddress(int fd) {
+  return Address::addressFromFd(fd);
+}
+
 Address::InstanceConstSharedPtr ListenerImpl::getOriginalDst(int fd) {
   return Utility::getOriginalDst(fd);
 }
 
 void ListenerImpl::listenCallback(evconnlistener*, evutil_socket_t fd, sockaddr* remote_addr,
                                   int remote_addr_len, void* arg) {
   ListenerImpl* listener = static_cast<ListenerImpl*>(arg);
-
   Address::InstanceConstSharedPtr final_local_address = listener->socket_.localAddress();
+
+  // Get the local address from the new socket if the listener is listening on the all hosts
+  // address (e.g., 0.0.0.0 for IPv4).
+  auto ip = final_local_address->ip();
+  if (ip && ip->isAnyAddress()) {
+    final_local_address = listener->getLocalAddress(fd);
+  }
+
   if (listener->options_.use_original_dst_ && final_local_address->type() == Address::Type::Ip) {
     Address::InstanceConstSharedPtr original_local_address = listener->getOriginalDst(fd);
-    if (original_local_address) {
-      final_local_address = original_local_address;
-    }
 
     // Hands off redirected connections (from iptables) to the listener associated with the
     // original destination address. If there is no listener associated with the original
     // destination address, the connection is handled by the listener that receives it.
-    // Note: A listener that has the use_original_dst flag set to true can still receive.
+    // Note: A listener that has the use_original_dst flag set to true can still receive
     // connections that are NOT redirected using iptables. If a connection was not redirected,
-    // the address returned by getOriginalDst() match the listener address. In this case the
-    // listener handles the connection directly and does not hand it off.
-    if (listener->socket_.localAddress() != final_local_address) {
+    // the address returned by getOriginalDst() matches the local address of the new socket.
+    // In this case the listener handles the connection directly and does not hand it off.
+    if (original_local_address && *original_local_address != *final_local_address) {
```

Comment:
nit: Add parens around `*original_local_address != *final_local_address` for clarity.